Group membership via google hd authentication

angusmcleod opened the pull request:
Update on #12446.

@davidtaylorhq A draft for an update on the associated group membership work. It's been rebased, brought in line with the latest changes and updated in the ways explained below. Let me know if you're onboard with the approach so far. I'll finish it off next weekend.

### Authentication and group retrieval

I've removed the secondary auth stuff, added a google-hd-specific boolean setting, and moved the group retrieval into a new google oauth2 omniauth strategy. I've made a new folder and added a spec for the strategy (something new in the codebase) as I thought it deserved a spec (which is also included). If we add in secondary auth to this strategy that too will need a spec. Some of the existing discourse omniauth strategies, e.g. Discord, may benefit from being seperated out and tested too.

### Group Provision and UI

As suggested, I've added a ``provides_groups?`` to the ``Auth::Authenticator``. Using this approach led to a new groups guardian method to use at the various group CRUD points. Speaking of which, I've added in support for group_associated_group creation in group creation (i.e. ``/g/custom/new``). Supporting it in both ``new`` and ``manage`` (update) also led to adding a new site attribute ``can_associate_groups``. I attempted to make this more group-route specific, however this leads to various clumsy workarounds to determine whether an admin can associate groups in ``/g/custom/new``. The site attribute approach proved the simplest and most performant (i.e. fewer ajax calls).

### Data model

I've made one change to the data model so far, namely to use ``after_commit`` hooks rather than ``after_create``/``after_destroy`` as using those hooks leads to duplicate record creation in ``Admin::GroupsController`` ``create``. I've yet to deal with the edge case you pointed out, i.e. multiple UserAssociatedGroups which link to the same group.

angusmcleod commented:
@davidtaylorhq this is ready for another review!

davidtaylorhq commented on app/models/group_associated_group.rb:
```diff
@@ -0,0 +1,53 @@
+  def remove_associated_users
+    DistributedMutex.synchronize("group_associated_group_#{group_id}_#{associated_group_id}") do
+      User.where("(
+        SELECT COUNT(user_id)
+        FROM user_associated_groups AS uag
+        WHERE uag.user_id = users.id
+        AND uag.associated_group_id IN (
+          SELECT associated_group_id FROM group_associated_groups AS gag
+          WHERE gag.group_id = ?
+        )
+      ) = 1", group.id).in_batches do |users|
```
See query suggestion in `UserAssociatedGroup.remove_associated_users`

davidtaylorhq replied:
```suggestion
      NOT EXISTS (
        SELECT 1
        FROM user_associated_groups uag
        JOIN group_associated_groups gag
        ON gag.associated_group_id = uag.associated_group_id
        WHERE uag.user_id = users.id
        AND gag.id != :gag_id
        AND gag.group_id = :group_id
      )", gag_id: id, group_id: group_id).in_batches do |users|
```

davidtaylorhq reviewed (commented):
This is looking great - thanks @angusmcleod. I've made a handful of suggestions, but nothing too major. I'd still like to take it for a spin locally before merging, but I think we're super close to having this done! 🚀

angusmcleod commented on app/models/user_associated_group.rb:
```diff
@@ -0,0 +1,46 @@
+  end
+
+  def remove_from_associated_groups
+    Group.where("(
+      SELECT COUNT(group_id)
+      FROM group_associated_groups AS gag
+      WHERE gag.group_id = groups.id
+      AND gag.associated_group_id IN (
+        SELECT associated_group_id FROM user_associated_groups AS uag
+        WHERE uag.user_id = ?
+      )
+    ) = 1", user_id).each do |group|
```
While I'm on board with making this more efficient, if you swap in your suggested SQL ``spec/models/user_associated_spec.rb`` won't pass. If you join ``gag`` and ``uag`` via ``associated_group_id`` you don't capture the records involving a different associated group granting membership to the same group.

davidtaylorhq replied:
It does join `gag` and `uag`, but it should still be able to pick up users from a different associated group. I think of it like "Look for all the UAGs (excluding the current UAG) which match a GAG which grants access to the given group"

I ran the spec locally and it looks like I had a couple of typos in my above suggestion. This version passes for me:

```suggestion
    Group.where("NOT EXISTS (
      SELECT 1
      FROM user_associated_groups uag
      JOIN group_associated_groups gag
      ON gag.associated_group_id = uag.associated_group_id
      WHERE uag.user_id = :user_id
      AND uag.id != :uag_id
      AND gag.group_id = groups.id
    )", uag_id: id, user_id: user_id).each do |group|
```

Ignore my comment about extracting the SQL into a constant. I now see that the use cases are slightly different. I've suggested a slightly amended query in `group_associated_groups.rb`

angusmcleod commented on app/services/group_action_logger.rb:
```diff
@@ -35,6 +35,26 @@ def log_remove_user_from_group(target_user)
   end
 
+  def log_add_user_to_group_automatically(target_user, subject)
+    @acting_user.id == Discourse::SYSTEM_USER_ID
+
```
Yes indeed. What a wierd thing for me to miss.

angusmcleod replied:
I've removed these new additional group action log methods entirely and just added an extra param to the existing method.

angusmcleod commented on app/services/group_action_logger.rb:
```diff
@@ -21,17 +21,19 @@ def log_remove_user_as_group_owner(target_user)
   end
 
-  def log_add_user_to_group(target_user)
+  def log_add_user_to_group(target_user, subject = nil)
```
@davidtaylorhq Just a note that I've revmoved those added methods (which seems to have deleted your comment) and added an additional param to the existing methods.

angusmcleod commented on lib/auth/google_oauth2_authenticator.rb:
```diff
@@ -35,8 +36,31 @@ def register_middleware(omniauth)
+    result
+  end
+
+  def provides_groups?
+    request_groups
+  end
+
+  protected
+
+  def request_groups
+    SiteSetting.google_oauth2_hd.present? && SiteSetting.google_oauth2_hd_groups
```
Yes. I think I might have made them seperate as I found that slightly more "descriptive", but happy to combine them.

angusmcleod commented on lib/auth/managed_authenticator.rb:
```diff
@@ -121,6 +121,8 @@ def after_create_account(user, auth)
     retrieve_avatar(user, association.info["image"])
     retrieve_profile(user, association.info)
+
+    auth.apply_associated_attributes! if auth.respond_to?(:apply_associated_attributes!)
```
Because some of the existing specs (not in this PR) use simple hashes instead of authtokens which causes an error here. But looking at it now, I should probably just update those existing tests.

angusmcleod commented on lib/auth/managed_authenticator.rb:
```diff
@@ -113,16 +113,16 @@ def after_authenticate(auth_token, existing_account: nil)
 
-  def after_create_account(user, auth)
-    auth_token = auth[:extra_data]
+  def after_create_account(user, auth_result)
```
@davidtaylorhq I've made this change to make it clear what this variable is (compare existing usage in spec for this class below).

angusmcleod commented:
@davidtaylorhq Sorry for the deluge of github emails I'm sure you got. I haven't used Github's "commit suggestion from reviewer" features before. While I like that it preserves the reviewer as one of the authors of the change, it does lead to various linting and minor fixes. In any event, I've incorporated your suggestions. The remaining failing test should be fixed once this is merged: https://github.com/discourse/discourse-oauth2-basic/pull/53

davidtaylorhq commented on spec/fabricators/associated_group_fabricator.rb:
```diff
@@ -0,0 +1,7 @@
+# frozen_string_literal: true
```
Please rename this file so it ends in `_spec.rb`

angusmcleod replied:
Note that this one is already ``_fabricator``.

davidtaylorhq commented on spec/models/group_associated_group.rb:
```diff
@@ -0,0 +1,41 @@
+# frozen_string_literal: true
```
Please rename this file so it ends in `_spec.rb`

davidtaylorhq commented on spec/models/user_associated_group.rb:
```diff
@@ -0,0 +1,41 @@
+# frozen_string_literal: true
```
Please rename this file so it ends in `_spec.rb`

davidtaylorhq commented on spec/requests/omniauth_callbacks_controller_spec.rb:
```diff
@@ -729,6 +729,67 @@ def enabled?
+          expect(user_associated_groups.length).to eq(2)
+          expect(user_associated_groups.exists?(associated_group_id: associated_groups.first.id)).to eq(true)
+          expect(user_associated_groups.exists?(associated_group_id: associated_groups.second.id)).to eq(true)
+        end
```
We should check that groups are removed correctly (right now, they aren't - see my other suggestions)
```suggestion

          # Second login, one group removed
          mock_omniauth_for_groups([group1])
          get "/auth/google_oauth2/callback.json", params: {
            scope: groups_scope.split(' '),
            code: 'abcde',
            hd: domain
          }
          expect(response.status).to eq(302)
          expect(UserAssociatedGroup.where(user_id: user.id).count).to eq(1)

          # Third login, all groups removed
          mock_omniauth_for_groups([])
          get "/auth/google_oauth2/callback.json", params: {
            scope: groups_scope.split(' '),
            code: 'abcde',
            hd: domain
          }
          expect(response.status).to eq(302)
          expect(UserAssociatedGroup.where(user_id: user.id).count).to eq(0)
        end
```

davidtaylorhq commented on lib/auth/result.rb:
```diff
@@ -95,6 +97,28 @@ def apply_user_attributes!
   end
 
+  def apply_associated_attributes!
+    if authenticator&.provides_groups? && associated_groups.present?
```
We should process an empty array, otherwise a user will never be 'removed' from their final group (see spec addition)
```suggestion
    if authenticator&.provides_groups? && !associated_groups.nil?
```

davidtaylorhq commented on lib/auth/google_oauth2_authenticator.rb:
```diff
@@ -35,8 +36,25 @@ def register_middleware(omniauth)
+
+  def after_authenticate(auth_token, existing_account: nil)
+    result = super
+    if provides_groups? && (groups = auth_token[:extra][:raw_groups]).any?
```
We should process an empty array, otherwise a user will never be 'removed' from their final group (see spec addition)
```suggestion
    if provides_groups? && (groups = auth_token[:extra][:raw_groups])
```

davidtaylorhq reviewed (commented):
I tried this out locally and it's sooooo smooth to use!

I added a few small suggestions, and responded to the comments regarding the SQL query.

>  Sorry for the deluge of github emails I'm sure you got. I haven't used Github's "commit suggestion from reviewer" features before. While I like that it preserves the reviewer as one of the authors of the change, it does lead to various linting and minor fixes.

Don't worry about the notifications - I have github emails turned off anyway ;) 

I think it's possible to batch together all the suggested changes and merge them in a single commit if you like. Or alternatively feel free to make the changes locally and push them up - I'm not bothered about the automatic shared authorship.

---

BTW, for easier testing/development, I've added group support to the `discourse-development-auth` plugin in this PR: https://github.com/discourse/discourse-development-auth/pull/1

---

We're planning a core beta release early next week. I think once that's done, we should be able to get this merged into `main` in the following days 🤞

burnoutberni commented:
Sorry for dropping in from the sidelines, but since you're mentioning the next release: do you think you'll be able to make this work with other non-Google Open ID providers as well by then or is this gonna take more time? Angus mentioned on Discourse Meta that this wouldn't be too difficult once the heavy-lifting is done.

In any case, thanks for all of your efforts!

angusmcleod commented:
@davidtaylorhq Thanks, will process this shortly.

@burnoutberni Note that David mentioned merging this _after_ the next beta. We'll only be including google groups in the initial release. This is a big change already and including other providers would be too much all in one go.

I too want other providers in there, but we can't rush a change like this.

angusmcleod commented:
@davidtaylorhq One thing we should consider is garbage collection of unused associated groups. Say for example that someone is setting up a google group structure: they name a group in google, log in, rename the group in google, log in...and do that a few times. You could end up with a lot of similar associated group names in the group admin UI, i.e. the dropdown could look something like

```
google_oauth2:my-cool-group
google_oauth2:my-coool-group
google_oauth2:my-coool-groups
```

I was thinking of handling that with a weekly cleanup job looking for "stale" associated groups, i.e. something like this on the AG model

```
def self.destroy_stale_groups!
    AssociatedGroup.joins("AS ag LEFT JOIN user_associated_groups AS uag ON ag.id = uag.associated_group_id")
      .joins("user_histories as uh ON uh.subject = (ag.name||':'||ag.provider_name||coalesce((':'||ag.provider_id), ''))")
      .where("
        uag.id IS NULL AND
        uh.updated_at < NOW() - INTERVAL '30 days';
      ").destroy_all
 end
```

This also raises the utiity of having the group ``label`` as the ``GroupHistory`` subject. It would be simpler to have the associated group id for a query like the above, however that would also mean less if that associated group was subsequently removed (i.e. the ``subject`` would then become useless). On balance I think it still makes sense to use the label as the ``subject``, however it would be nice to have an ``associated_group_id`` column in ``group_histories``. Do you think it's worth adding one?

Moreover, looking at the ``label`` again now, I'm not sure the difference between the ``label`` on the server (name:provider_name:provider_id) and the client (provider_name:name) makes sense. I think my original thinking behind that was something like the id would be messy on the client, but now I'm not sure. Two issues to resolve:

1. the (name:provider_name) order should be consistent. What order would you prefer?
2. do you think the non-inclusion of the ``provider_id`` on the client makes sense?

davidtaylorhq commented:
Regarding garbage-collecting old AssociatedGroups, I think we should only do it automatically if

1. The AssociatedGroup has 0 UserAssociatedGroups **and 0 GroupAssociatedGroups**
2. It was 'last seen' more than 1 week ago (or similar)

For (2), I don't think we should join on the `user_histories` table. That could be extremely slow on large sites. Instead we could add a `last_seen` column to AssociatedGroups (similar to the `last_seen` column on UserAssociatedAccounts). That column can be bumped every time we receive a response from the identity provider which contains that group. Then the query to lookup stale groups should be something like

```ruby
AssociatedGroup.left_joins(:group_associated_groups, :user_associated_groups)
               .where("group_associated_groups.id IS NULL AND user_associated_groups.id IS NULL")
               .where("last_seen_at < ?", 1.week.ago)
```
---

For the other questions:

1. I think I prefer `provider_name:name`, so that an alphabetical list will group by provider. But if you prefer otherwise, I don't mind too much
2. I think I prefer excluding it. If we do include it, it should be the last item, and probably in brackets to avoid confusion. I imagine some providers could have very long (e.g. >20 char) group IDs. Again, I don't mind too much either way. We can always add/remove it later based on user feedback

burnoutberni commented:
@angusmcleod Totally understand, thanks for the update <3

angusmcleod commented:
@davidtaylorhq Ok, I've added [garbage collection](https://github.com/discourse/discourse/pull/14835/commits/29731ffe4374413c6982d114553ddda5c5e64f57), and addressed the [label issue](https://github.com/discourse/discourse/pull/14835/commits/e204c6cb2e741d57364784b427a1ea9e6f720042). Ready for another look.

davidtaylorhq commented:
Looks great, thanks @angusmcleod - I'll aim to get this merged in the next couple of weeks 🚀

davidtaylorhq commented on config/locales/server.en.yml:
```diff
@@ -1729,6 +1729,7 @@ en:
     google_oauth2_client_secret: "Client secret of your Google application."
     google_oauth2_prompt: "An optional space-delimited list of string values that specifies whether the authorization server prompts the user for reauthentication and consent. See <a href='https://developers.google.com/identity/protocols/OpenIDConnect#prompt' target='_blank'>https://developers.google.com/identity/protocols/OpenIDConnect#prompt</a> for the possible values."
     google_oauth2_hd: "An optional Google Apps Hosted domain that the sign-in will be limited to. See <a href='https://developers.google.com/identity/protocols/OpenIDConnect#hd-param' target='_blank'>https://developers.google.com/identity/protocols/OpenIDConnect#hd-param</a> for more details."
+    google_oauth2_hd_groups: "Retrieve users' Google groups on the hosted domain on authentication. Retrieved Google groups can be used to grant automatic Discourse group membership (see group settings)."
```
```suggestion
    google_oauth2_hd_groups: "(experimental) Retrieve users' Google groups on the hosted domain on authentication. Retrieved Google groups can be used to grant automatic Discourse group membership (see group settings)."
```
